In [1]:
import pandas as pd

# CustomerDemographic data quality assessment

In [2]:
scpl_data_Customer_Demographic  =  pd.read_excel('/Users/zhenliu15471/code/zliu15471/KPMG/raw_data/KPMG_VI_New_raw_data_update_final.xlsx','CustomerDemographic')

In [3]:
scpl_data_Customer_Demographic.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                        customer_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  3                                                                                                                              
4                                                  4                                                                                                                              

   Unnamed: 1 Unnamed: 2 Unnamed: 3                           Unnamed: 4  \
0  first_name  last_name     gender  past_3_years_bike_related_purchases   
1     Laraine  Medendorp          F                                   93   
2         Eli    Bockman       Male                                   81   
3       Arlin     Dearle       Male                                   61   
4      Talbot        NaN       Male                                   33   

            Unnamed: 5              Unnamed: 6             Unnamed: 7  \
0                  DOB               job_title  job_industry_category   
1  1953-10-12 00:00:00     Executive Secretary                 Health   
2  1980-12-16 00:00:00  Administrative Officer     Financial Services   
3  1954-01-20 00:00:00      Recruiting Manager               Property   
4  1961-10-03 00:00:00                     NaN                     IT   

       Unnamed: 8          Unnamed: 9  \
0  wealth_segment  deceased_indicator   
1   Mass Customer                   N   
2   Mass Customer                   N   
3   Mass Customer                   N   
4   Mass Customer                   N   

                                         Unnamed: 10 Unnamed: 11 Unnamed: 12  
0                                            default    owns_car      tenure  
1                                                 "'         Yes          11  
2                       <script>alert('hi')</script>         Yes          16  
3                                2018-02-01 00:00:00         Yes          15  
4  () { _; } >_[$($())] { touch /tmp/blns.shellsh...          No           7

## corrrect values (accuracy)

In [4]:
headers = scpl_data_Customer_Demographic.iloc[0]
new_df  = pd.DataFrame(scpl_data_Customer_Demographic.values[1:], columns=headers)

/var/folders/4y/7vss76ld1rj1ybqx7xpskc5m0000gn/T/ipykernel_45118/2244132357.py:2: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype={value.dtype})
  new_df  = pd.DataFrame(scpl_data_Customer_Demographic.values[1:], columns=headers)


In [5]:
new_df.head()

0 customer_id      first_name  last_name  gender  \
0           1         Laraine  Medendorp       F   
1           2             Eli    Bockman    Male   
2           3           Arlin     Dearle    Male   
3           4          Talbot        NaN    Male   
4           5  Sheila-kathryn     Calton  Female   

0 past_3_years_bike_related_purchases        DOB               job_title  \
0                                  93 1953-10-12     Executive Secretary   
1                                  81 1980-12-16  Administrative Officer   
2                                  61 1954-01-20      Recruiting Manager   
3                                  33 1961-10-03                     NaN   
4                                  56 1977-05-13           Senior Editor   

0 job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

0                                            default owns_car tenure  
0                                                 "'      Yes     11  
1                       <script>alert('hi')</script>      Yes     16  
2                                2018-02-01 00:00:00      Yes     15  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No      7  
4                                                NIL      Yes      8

In [6]:
new_df.shape

(4000, 13)

In [7]:
new_df.describe()

/var/folders/4y/7vss76ld1rj1ybqx7xpskc5m0000gn/T/ipykernel_45118/3143706273.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  new_df.describe()


0       customer_id first_name last_name  gender  \
count        4000.0       4000      3875    4000   
unique       4000.0       3139      3725       6   
top             1.0        Max  Pristnor  Female   
freq            1.0          5         3    2037   
first           NaN        NaN       NaN     NaN   
last            NaN        NaN       NaN     NaN   

0       past_3_years_bike_related_purchases                  DOB  \
count                                4000.0                 3913   
unique                                100.0                 3448   
top                                    16.0  1978-01-30 00:00:00   
freq                                   56.0                    7   
first                                   NaN  1843-12-21 00:00:00   
last                                    NaN  2002-03-11 00:00:00   

0                                  job_title job_industry_category  \
count                                   3494                  3344   
unique                                   195                     9   
top     Business Systems Development Analyst         Manufacturing   
freq                                      45                   799   
first                                    NaN                   NaN   
last                                     NaN                   NaN   

0      wealth_segment deceased_indicator  default owns_car  tenure  
count            4000               4000   3698.0     4000  3913.0  
unique              3                  2     90.0        2    22.0  
top     Mass Customer                  N    100.0      Yes     7.0  
freq             2000               3998    113.0     2024   235.0  
first             NaN                NaN      NaN      NaN     NaN  
last              NaN                NaN      NaN      NaN     NaN

In [8]:
new_df.dtypes.value_counts()

object            12
datetime64[ns]     1
dtype: int64

In [9]:
new_df.nunique()

0
customer_id                            4000
first_name                             3139
last_name                              3725
gender                                    6
past_3_years_bike_related_purchases     100
DOB                                    3448
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
default                                  90
owns_car                                  2
tenure                                   22
dtype: int64

In [25]:
new_df.duplicated().sum()

0

0

In [15]:
new_df[new_df['DOB']< '1942']

0    customer_id first_name last_name  gender  \
33            34   Jephthah  Bachmann       U   
719          720     Darrel     Canet    Male   
1091        1092     Katlin   Creddon  Female   
3409        3410    Merrili   Brittin  Female   

0    past_3_years_bike_related_purchases        DOB           job_title  \
33                                    59 1843-12-21     Legal Assistant   
719                                   67 1931-10-23  Recruiting Manager   
1091                                  56 1935-08-22  VP Quality Control   
3409                                  93 1940-09-22                 NaN   

0    job_industry_category     wealth_segment deceased_indicator  \
33                      IT  Affluent Customer                  N   
719                 Retail  Affluent Customer                  N   
1091                Retail      Mass Customer                  N   
3409              Property      Mass Customer                  N   

0                default owns_car tenure  
33                   NaN       No     20  
719   1;DROP TABLE users       No      6  
1091  ì¬íê³¼íì ì´íì°êµ¬ì       No      5  
3409                   á       No     16

In [17]:
new_df.gender.unique()

array(['F', 'Male', 'Female', 'U', 'Femal', 'M'], dtype=object)

### DOB:  there is a person born in 1843, which is clearly wrong. should remove the data.
### gender: a few errors occur in the gender like 'F', 'Femal','M'. these should be replaced by 'Male' and 'Female'
### default: the default values are messy code, should drop them.


## data fields with values (completeness)

In [18]:
(new_df.isnull().sum()/len(new_df))*100

0
customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
default                                 7.550
owns_car                                0.000
tenure                                  2.175
dtype: float64

### there are several columns which have null values. Among them, the job_title and job_industry_category arrive 12.65% and 16.4%
### all these null values can be filled by simple inputer.

## values free from contradiction (consistency)

### the default column don't have consistency, should remove 

In [19]:
new_df.default.unique()

array(['"\'', "<script>alert('hi')</script>",
       datetime.datetime(2018, 2, 1, 0, 0),
       '() { _; } >_[$($())] { touch /tmp/blns.shellshock2.fail; }',
       'NIL', 'ðµ ð ð ð', 'â°â´âµâââ', '(â¯Â°â¡Â°ï¼â¯ï¸µ â»ââ»)', '0/0',
       'ð©ð½', 'Åâ´®â\xa0¥¨ËøÏââ', 'nil', -100, 'â°â´âµ', 'ð', 1e+96,
       'ï¾ï½¥â¿ã¾â²(｡ââ¿â｡)â±â¿ï½¥ï¾', 'Ωâçââ«Ëµâ¤â¥Ã·',
       'Åâ´â°ËèËÃâââ', 'ï¼ï¼ï¼',
       '../../../../../../../../../../../etc/hosts',
       '×Ö¸×ְתָ×testاÙصÙØ\xadات اÙتÙØ\xadÙÙ', '<>?:"{}|_+',
       '\'\'\'\'"', ",./;'[]\\-=",
       '() { 0; }; touch /tmp/blns.shellshock1.fail;',
       'ì¬íê³¼íì ì´íì°êµ¬ì', 'testâ\xa0testâ«',
       '0ï¸â£ 1ï¸â£ 2ï¸â£ 3ï¸â£ 4ï¸â£ 5ï¸â£ 6ï¸â£ 7ï¸â£ 8ï¸â£ 9ï¸â£ ð',
       nan, '!@#$%^&*()', "'",
       '̦HÍ̬̤Ì̤eÍ ÍÌÌ¥ÌÌ»ÍÌwÌhÌ̯ÍoÌÍÌÍ̱̮ Ò̺ÌÌÌÍW̷̼Ì\xada̺̪ÍįÍÍÌ\xadÍ̯Ìt̶̼̮sÌÌÍÍ Ì\xa0Ì«Ì\xa0BÌ»ÍÍÍÍ̳e̵h̵̬ÍÌ«ÍiÌ̹Í̳̳̮ÍÌ«nÍd̴̪ÌÌ ÍÌ°ÍÌ©ÍÍÍ̲TÍ¢Í̼Í̪hÍÍ̮̻e̬ÌÍÌ Ì¤Ì¹ÌWÍÍÍÌÌÍÍaÍÍÍ̹̼',
       '../../../

## values up to date (currency)

### there is no problem for that

## data items with value Meta-data (relevancy)

In [20]:
new_df.wealth_segment.unique()

array(['Mass Customer', 'Affluent Customer', 'High Net Worth'],
      dtype=object)

In [21]:
new_df.job_title.unique()

array(['Executive Secretary', 'Administrative Officer',
       'Recruiting Manager', nan, 'Senior Editor', 'Media Manager I',
       'Business Systems Development Analyst', 'Senior Quality Engineer',
       'Nuclear Power Engineer', 'Developer I', 'Account Executive',
       'Junior Executive', 'Media Manager IV', 'Sales Associate',
       'Professor', 'Geological Engineer', 'Project Manager',
       'Safety Technician I', 'Research Assistant I',
       'Accounting Assistant III', 'Editor', 'Research Nurse',
       'Safety Technician III', 'Staff Accountant III', 'Legal Assistant',
       'Product Engineer', 'Information Systems Manager',
       'VP Quality Control', 'Social Worker', 'Senior Cost Accountant',
       'Assistant Media Planner', 'Payment Adjustment Coordinator',
       'Food Chemist', 'Accountant III', 'Director of Sales',
       'Senior Financial Analyst', 'Registered Nurse',
       'Biostatistician II', 'Computer Systems Analyst II',
       'Software Test Engineer II', 

### the first name and last name should not be relevant with the bikes & cycling situation, should be dropped.
### the job title column has too many unique values and is hard to be quantified, should be dropped.

## data containing allowable values (validity)

### default values

### gender

## records that are duplicated (uniqueness)

In [22]:
new_df.nunique()

0
customer_id                            4000
first_name                             3139
last_name                              3725
gender                                    6
past_3_years_bike_related_purchases     100
DOB                                    3448
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
default                                  90
owns_car                                  2
tenure                                   22
dtype: int64

### the uniqueness of values are fine

In [26]:
new_df.duplicated().sum()

0

# Customer Addresses data quality assessment

In [30]:
scpl_data_CustomerAddress  =  pd.read_excel('/Users/zhenliu15471/code/zliu15471/KPMG/raw_data/KPMG_VI_New_raw_data_update_final.xlsx','CustomerAddress')

In [31]:
scpl_data_CustomerAddress.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                        customer_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  4                                                                                                                              
4                                                  5                                                                                                                              

            Unnamed: 1 Unnamed: 2       Unnamed: 3 Unnamed: 4  \
0              address   postcode            state    country   
1   060 Morning Avenue       2016  New South Wales  Australia   
2  6 Meadow Vale Court       2153  New South Wales  Australia   
3   0 Holy Cross Court       4211              QLD  Australia   
4  17979 Del Mar Point       2448  New South Wales  Australia   

           Unnamed: 5  
0  property_valuation  
1                  10  
2                  10  
3                   9  
4                   4

In [32]:
headers = scpl_data_CustomerAddress.iloc[0]
new_df_a  = pd.DataFrame(scpl_data_CustomerAddress.values[1:], columns=headers)

In [33]:
new_df_a.head()

0 customer_id              address postcode            state    country  \
0           1   060 Morning Avenue     2016  New South Wales  Australia   
1           2  6 Meadow Vale Court     2153  New South Wales  Australia   
2           4   0 Holy Cross Court     4211              QLD  Australia   
3           5  17979 Del Mar Point     2448  New South Wales  Australia   
4           6     9 Oakridge Court     3216              VIC  Australia   

0 property_valuation  
0                 10  
1                 10  
2                  9  
3                  4  
4                  9

In [34]:
new_df_a.dtypes.value_counts()

object    6
dtype: int64

In [35]:
new_df_a.nunique()

0
customer_id           3999
address               3996
postcode               873
state                    5
country                  1
property_valuation      12
dtype: int64

In [36]:
new_df_a.duplicated().sum()

0

In [37]:
(new_df_a.isnull().sum()/len(new_df_a))*100

0
customer_id           0.0
address               0.0
postcode              0.0
state                 0.0
country               0.0
property_valuation    0.0
dtype: float64

In [40]:
new_df_a.state.unique()

array(['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria'], dtype=object)

## for the state column, the 'VIC' and 'Victoria', 'New South Wales' and 'NSW' belong to same values and should be replaced by 'VIC' and 'NSW'
## for the address column, as the postcode is shown to represent the information of location and the values of address are hard to be quantified, the address column should be dropped.
## the country column is only one value and can't reflect any correlation, should be dropped.

# Transactions 

In [41]:
scpl_data_transactions  =  pd.read_excel('/Users/zhenliu15471/code/zliu15471/KPMG/raw_data/KPMG_VI_New_raw_data_update_final.xlsx','Transactions')

In [42]:
scpl_data_transactions.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                     transaction_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  3                                                                                                                              
4                                                  4                                                                                                                              

   Unnamed: 1   Unnamed: 2           Unnamed: 3    Unnamed: 4    Unnamed: 5  \
0  product_id  customer_id     transaction_date  online_order  order_status   
1           2         2950  2017-02-25 00:00:00         False      Approved   
2           3         3120  2017-05-21 00:00:00          True      Approved   
3          37          402  2017-10-16 00:00:00         False      Approved   
4          88         3135  2017-08-31 00:00:00         False      Approved   

       Unnamed: 6    Unnamed: 7     Unnamed: 8    Unnamed: 9 Unnamed: 10  \
0           brand  product_line  product_class  product_size  list_price   
1           Solex      Standard         medium        medium       71.49   
2   Trek Bicycles      Standard         medium         large     2091.47   
3      OHM Cycles      Standard            low        medium     1793.43   
4  Norco Bicycles      Standard         medium        medium     1198.46   

     Unnamed: 11              Unnamed: 12  
0  standard_cost  product_first_sold_date  
1          53.62                    41245  
2         388.92                    41701  
3         248.82                    36361  
4          381.1                    36145

In [43]:
headers = scpl_data_transactions.iloc[0]
new_df_t  = pd.DataFrame(scpl_data_transactions.values[1:], columns=headers)

In [44]:
new_df_t

0     transaction_id product_id customer_id transaction_date online_order  \
0                  1          2        2950       2017-02-25        False   
1                  2          3        3120       2017-05-21         True   
2                  3         37         402       2017-10-16        False   
3                  4         88        3135       2017-08-31        False   
4                  5         78         787       2017-10-01         True   
...              ...        ...         ...              ...          ...   
19995          19996         51        1018       2017-06-24         True   
19996          19997         41         127       2017-11-09         True   
19997          19998         87        2284       2017-04-14         True   
19998          19999          6        2764       2017-07-03        False   
19999          20000         11        1144       2017-09-22         True   

0     order_status           brand product_line product_class product_size  \
0         Approved           Solex     Standard        medium       medium   
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19995     Approved      OHM Cycles     Standard          high       medium   
19996     Approved           Solex         Road        medium       medium   
19997     Approved      OHM Cycles     Standard        medium       medium   
19998     Approved      OHM Cycles     Standard          high       medium   
19999     Approved   Trek Bicycles     Standard        medium        small   

0     list_price standard_cost product_first_sold_date  
0          71.49         53.62                   41245  
1        2091.47        388.92                   41701  
2        1793.43        248.82                   36361  
3        1198.46         381.1                   36145  
4         1765.3        709.48                   42226  
...          ...           ...                     ...  
19995    2005.66        1203.4                   37823  
19996     416.98        312.74                   35560  
19997     1636.9         44.71                   40410  
19998     227.88        136.73                   38216  
19999    1775.81       1580.47                   36334  

[20000 rows x 13 columns]

In [46]:
new_df_t.dtypes.value_counts()

object            12
datetime64[ns]     1
dtype: int64

In [47]:
new_df_t.nunique()

0
transaction_id             20000
product_id                   101
customer_id                 3494
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   296
standard_cost                103
product_first_sold_date      100
dtype: int64

In [50]:
(new_df_t.isnull().sum()/len(new_df_t))*100

0
transaction_id             0.000
product_id                 0.000
customer_id                0.000
transaction_date           0.000
online_order               1.800
order_status               0.000
brand                      0.985
product_line               0.985
product_class              0.985
product_size               0.985
list_price                 0.000
standard_cost              0.985
product_first_sold_date    0.985
dtype: float64

In [53]:
new_df_t.transaction_date.describe()

/var/folders/4y/7vss76ld1rj1ybqx7xpskc5m0000gn/T/ipykernel_45118/1215687526.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  new_df_t.transaction_date.describe()


count                   20000
unique                    364
top       2017-02-14 00:00:00
freq                       82
first     2017-01-01 00:00:00
last      2017-12-30 00:00:00
Name: transaction_date, dtype: object

## the transaction_date contain values of one year, not just latest three month. should filter the dates
## confused about the product_first_sold_date because these values are not date just numbers.
## there are still some missing values in the columns and only online_order is over 1%, the missing values can be filled 
